File: libraries/node/src/parser/lex.ts
import { UrcError } from "../error.js";
import type { Meter } from "../model.js";
import { SYNTAX, rule } from "../strings.js";

const INT_RE = /^-?\d+$/;
const FLOAT_RE = /^-?\d+(\.\d+)?$/;
const UINT_RE = /^\d+$/;
const METER_RE = /^(\d+)\/(\d+)$/;
const TYPE_RE = /^(\d+)(?:\+(\d+))?$/;

/**
 * Checks that a token is entirely ASCII digits.
 */
export function validUint(token: string): boolean {
	return UINT_RE.test(token);
}

/**
 * Parses a signed decimal integer token.
 */
export function intValue(token: string, line: number): number {
	if (!INT_RE.test(token)) 
		throw new UrcError(SYNTAX, line, `invalid integer: '${token}'`);
	
	return Number(token);
}

/**
 * Parses a plain decimal float token (no exponent notation).
 */
export function floatValue(token: string, line: number): number {
	if (!FLOAT_RE.test(token)) 
		throw new UrcError(SYNTAX, line, `invalid float: '${token}'`);
	
	return Number(token);
}

/**
 * Parses a comma-separated float list; rejects empty items.
 */
export function floatList(value: string, line: number): number[] {
	const values: number[] = [];
	for (const raw of value.split(",")) {
		const token = raw.trim();
		if (token === "") 
			throw new UrcError(SYNTAX, line, "empty value in list");
		
		values.push(floatValue(token, line));
	}
	return values;
}

/**
 * Parses a `keys` or `keys+special` Type value into its pair.
 */
export function layoutType(value: string, line: number): [number, number] {
	const match = TYPE_RE.exec(value);
	if (match === null) 
		throw new UrcError(SYNTAX, line, `invalid Type value: '${value}'`);
	
	const keys = Number(match[1]);
	const hadSpecial = match[2] !== undefined;
	const special = hadSpecial ? Number(match[2]) : 0;
	if (keys < 1 || (hadSpecial && special < 1)) 
		throw new UrcError(SYNTAX, line, "Type values must be positive");
	
	return [keys, special];
}

/**
 * Parses a `beats/noteValue` meter token.
 */
export function meter(token: string, line: number): Meter {
	const match = METER_RE.exec(token);
	if (match === null) 
		throw new UrcError(rule(17), line, `invalid meter: '${token}'`);
	
	const beats = Number(match[1]);
	const noteValue = Number(match[2]);
	if (beats < 1 || noteValue < 1) 
		throw new UrcError(rule(17), line, `invalid meter: '${token}'`);
	
	return { beats, noteValue };
}

File: libraries/rust/src/parser/lex.rs
//! Scalar and list token parsing for URC field values.

use crate::error::{Result, UrcError};
use crate::model::Meter;
use crate::strings::{SYNTAX, rule};

pub(super) fn valid_uint(token: &str) -> bool {
    !token.is_empty() && token.bytes().all(|b| b.is_ascii_digit())
}

pub(super) fn parse_int(token: &str, line: u32) -> Result<i64> {
    let body = token.strip_prefix('-').unwrap_or(token);
    if !valid_uint(body) {
        return Err(UrcError::new(
            SYNTAX,
            line,
            format!("invalid integer: {token:?}"),
        ));
    }
    token
        .parse::<i64>()
        .map_err(|_| UrcError::new(SYNTAX, line, format!("invalid integer: {token:?}")))
}

pub(super) fn valid_float(token: &str) -> bool {
    let body = token.strip_prefix('-').unwrap_or(token);
    match body.split_once('.') {
        Some((int_part, frac)) => valid_uint(int_part) && valid_uint(frac),
        None => valid_uint(body),
    }
}

pub(super) fn parse_float(token: &str, line: u32) -> Result<f64> {
    if !valid_float(token) {
        return Err(UrcError::new(
            SYNTAX,
            line,
            format!("invalid float: {token:?}"),
        ));
    }
    token
        .parse::<f64>()
        .map_err(|_| UrcError::new(SYNTAX, line, format!("invalid float: {token:?}")))
}

pub(super) fn float_list(value: &str, line: u32) -> Result<Vec<f64>> {
    value
        .split(',')
        .map(|token| parse_float(token.trim(), line))
        .collect()
}

pub(super) fn layout_type(value: &str, line: u32) -> Result<(u64, u64)> {
    let malformed = || UrcError::new(SYNTAX, line, format!("invalid Type value: {value:?}"));
    let (keys, special) = match value.split_once('+') {
        Some((keys, special)) => (keys, Some(special)),
        None => (value, None),
    };
    if !valid_uint(keys) {
        return Err(malformed());
    }
    let had_special = special.is_some();
    let special = match special {
        None => 0,
        Some(text) if valid_uint(text) => text.parse::<u64>().unwrap_or(u64::MAX),
        Some(_) => return Err(malformed()),
    };
    let keys = keys.parse::<u64>().unwrap_or(u64::MAX);
    if keys < 1 || (had_special && special < 1) {
        return Err(UrcError::new(SYNTAX, line, "Type values must be positive"));
    }
    Ok((keys, special))
}

pub(super) fn parse_meter(token: &str, line: u32) -> Result<Meter> {
    let invalid = || UrcError::new(rule(17), line, format!("invalid meter: {token:?}"));
    let Some((beats, note_value)) = token.split_once('/') else {
        return Err(invalid());
    };
    if !valid_uint(beats) || !valid_uint(note_value) {
        return Err(invalid());
    }
    let beats = beats.parse::<u64>().unwrap_or(u64::MAX);
    let note_value = note_value.parse::<u64>().unwrap_or(u64::MAX);
    if beats < 1 || note_value < 1 {
        return Err(invalid());
    }
    Ok(Meter { beats, note_value })
}
